### DISCUSSION TITLE
Native recovery flow error with code strategy

### DISCUSSION THREAD

**Discussion by NiteArie** [ory/kratos#2959] [Q&A] [answered]
Hi there,

I'm currently using the new code strategy for account recovery. I confirmed that I have upgraded to using Kratos v0.11 with the new v1.1 JavaScript SDK.

I'm initializing the flow using `createNativeRecoveryFlow` and successfully in getting the code from the email after using the method `updateRecoveryFlow`. 

The problem here is I can't complete the recovery flow with the code that I got from the email, on submission I called 
```js
updateRecoveryFlow({
      flow: flow.id,
      updateRecoveryFlowBody: {
        method: 'code',
        email,
        code,
      },
    })
``` 
With the `flow.id` being the flow that was initialized before
I got back this error insteading of completion:

```js
{
 error: {
    id: 'browser_location_change_required',
    code: 422,
    status: 'Unprocessable Entity',
    reason: 'In order to complete this flow please redirect the browser to: http://127.0.0.1:3000/recovery?flow=fb058b5f-43d3-435d-b5d0-bd01a93968ac',
    message: 'browser location change required'
  },
  redirect_browser_to: 'http://127.0.0.1:3000/recovery?flow=fb058b5f-43d3-435d-b5d0-bd01a93968ac'
}
```

Am I doing something wrong in the process of recovering or this isn't supported? Thanks for the answer in advance.

**Comment by nicolasburtey** [comment]
I'm having the same issue

**Reply by nicolasburtey** [reply]
it seems the email should not be here? https://github.com/ory/kratos/discussions/2923

**Reply by NiteArie** [reply]
@nicolasburtey Yeah removing the `email` from the body doesn't solve the issue. The error is still the same.

**Reply by vinckr** [reply]
Can you please confirm it is also not this: 

> Copying the recovery code from the email included a leading space. After trimming the input, the account recovery was successful.

**Reply by NiteArie** [reply]
@vinckr Yes. I have thoroughly check my code and read all the issues/discussions related to the recovery process to not make this mistake.

**Comment by micheljung** [comment]
To reproduce:

```
domain=example.com

# Init flow
actionUrl=$(curl -s "https://$domain/self-service/recovery/api" | jq -r .ui.action)

# Request code
curl -X POST -H "Content-Type: application/json" -d '{"email": "user@example.com", "method": "code"}' "$actionUrl"

# Submit code (added 'Accept: application/json' according to https://www.ory.sh/docs/reference/api#tag/frontend/operation/updateRecoveryFlow)
curl -i -X POST -H "Content-Type: application/json" -H "Accept: application/json" -d '{"code": "467404", "method": "code"}' "$actionUrl"
```

Response:
```
HTTP/2 422
cache-control: private, no-cache, no-store, must-revalidate
content-type: application/json
date: Thu, 29 Dec 2022 11:36:50 GMT
set-cookie: csrf_token_5c349d55abcda945dde7b96dc14a4105480ede67710fc0f844c0b2ab2b2ff947=LaacqKcqU+nCUdC9mylMGbpy4fYv6KATMUlLzS0W5D4=; Path=/; Domain=REDACTED; Max-Age=31536000; HttpOnly; Secure; SameSite=Lax
set-cookie: ory_kratos_session=MTY3MjMxMzgxMHxSRnNoYU43UVVNRXI4NnJmb000V0JsYVBKTXZxOEZvbUhTZ3JWVWM1ZWVwVU45Rm5rbzB3dmJMZGExR0J5UkNpX0xkMWtyc2dKSXdMbWNFUmlmYndGbGVXeGUtbm5LTXJ1RnJUODVDdTRINC1ta19EbFB6TXFNYVlnbG9JeGpTZDdfWGtxak9RXzZidEZ5U01lZHBjcVh0ZUU1YTJPdHZnVXRZNVF3bnQwVVNEM0VkdHd2QU9iSmRSNTBYemdZMHpFdFY0SW1NaWY0eV8yVnVZTEh0U052STBLSEpLdnktNHdxV1hac1NQdkNUZFFNR29oazlJZVpPS1NHWTIxT1M4REc2UXBJR2hYVnJ1fDKie2-H449tft8Pxflm39sJjQg3yQU-7mEzk0znB749; Path=/; Domain=REDACTED; Expires=Fri, 30 Dec 2022 11:36:49 GMT; Max-Age=86399; HttpOnly; Secure; SameSite=Lax
vary: Cookie
content-length: 414

{"error":{"id":"browser_location_change_required","code":422,"status":"Unprocessable Entity","reason":"In order to complete this flow please redirect the browser to: https://REDACTED/settings?flow=fa8b0aaa-5ffe-445b-9ef8-d7e594c8a63b","message":"browser location change required"},"redirect_browser_to":"https://REDACTED/settings?flow=fa8b0aaa-5ffe-445b-9ef8-d7e594c8a63b"}
```

**Comment by staklau** [comment]
New Ory Network user here, and I'm having the same issue but with `createBrowserRecoveryFlow`.

**Comment by jonas-jonas** [comment]
Hi all! 

Sorry for the confusion on this topic. The current state of the recovery flow does not support native apps. There are multiple reasons for this:

- There currently is no good way (in Kratos) for redirecting users to a different flow (type), but the current implementation requires the use of the "Settings" flow. That's why you're seeing the `browser_location_change_required` error. There is an open issue in which we track the implementation of that https://github.com/ory/kratos/issues/2884
- On the other hand, the current implementation of the recovery flow requiring the "Settings" flow is not optimal, and there have been discussions on whether to make the Recovery flow "self-contained" by allowing users to re-set their password entirely inside the flow (making the requirement to redirect to another flow obsolete), but that requires a thorough technical design as it needs to be sufficiently secure!

Lastly, the documentation on the recovery flow has just been reworked, but we failed to include any warnings on this incompatibility, sorry about that.

As a workaround, you can implement the recovery flow using a browser window in your app. We know that this is not optimal, and have been thinking about how to solve it for a while, but are also busy with other, pressing things.

**Reply by NiteArie** [reply]
Thank you so much

**Reply by mk1020** [reply]
hey @jonas-jonas ! Our company is migrating from auth0 to ory, we are using react native and faced this problem. Is there any updates or solutions for this problem, because this is a very serious problem and it's been a year since your reply

**Reply by vinckr** [reply]
Hello @mk1020 
Are you using Ory Network or self-hosting Ory Kratos? Feel free to reach out to me at vincent@ory.sh if you want to chat about your use case.

**Reply by mk1020** [reply]
@vinckr we're using Ory Network. Thank you!

**Reply by mk1020** [reply]
We are using react native and we are facing a problem where we cannot finish the password recovery flow. I get a code emailed to me but when I enter the code, instead of a successful response I get a browser_location_change_required error. I have tried different ways but nothing works, i also tried using webview but the user experience is very bad. Also, you have a repository called kratos-selfservice-ui-react-native, but password recovery doesn't work there either.

**Reply by nicolasburtey** [reply]
facing same problem here

**Reply by nicolasburtey** [reply]
in case you were able to solve the problem, let me know @mk1020

**Reply by jonas-jonas** [reply]
Please see https://github.com/ory/kratos/discussions/2959#discussioncomment-9164181 for an updated answer.

**Comment by nicolasburtey** [comment]
actually I made it working with kratos v1 but can't make it work with kratos v1.1

I guess we'll wait until `createNativeRecoveryFlow` is properly implemented

**Reply by jonas-jonas** [reply]
Please see https://github.com/ory/kratos/discussions/2959#discussioncomment-9164181

**Comment by jonas-jonas** [comment, marked as answer]
Some time has passed since my initial comment (https://github.com/ory/kratos/discussions/2959#discussioncomment-4643625).
 
We have since implemented a fix for the issue here https://github.com/ory/kratos/pull/3624. The way to enable it, is to add the following to your config:

```yaml
feature_flags:
  use_continue_with_transitions: true
```

After setting that, instead of returning a `browser_location_change_required` error, Kratos should return a reference to the settings flow, that the user should then use to update their password.
